src/broadcast/scrobblingmanager.cpp

```diff
@@ -0,0 +1,210 @@
+        QMutexLocker locker(&m_mutex);
+        for (TrackInfo *trackInfo : m_trackList) {
+            if (trackInfo->m_pTrack == pResumedTrack) {
+                trackInfo->m_trackInfo.resumePlayedTime();
+                break;
+            }
+        }  
+    }
+}
+
+void ScrobblingManager::slotLoadingTrack(TrackPointer pNewTrack, TrackPointer pOldTrack) {
+    BaseTrackPlayer *sourcePlayer = 
+        qobject_cast<BaseTrackPlayer*>(sender());
+    DEBUG_ASSERT(sourcePlayer);
+    if (pOldTrack) {
+        m_tracksToBeReset.append(TrackToBeReset(pOldTrack,
+            sourcePlayer->getGroup()));
+    }
+}
+
+void ScrobblingManager::slotNewTrackLoaded(TrackPointer pNewTrack) {
+    //Empty player gives a null pointer.
+    if (!pNewTrack)
+        return;          
+    BaseTrackPlayer *player = qobject_cast<BaseTrackPlayer*>(sender());    
+    DEBUG_ASSERT(player);
+    QMutexLocker locker(&m_mutex);    
+    bool trackAlreadyAdded = false;
+    for (TrackInfo *trackInfo : m_trackList) {
+        if (trackInfo->m_pTrack == pNewTrack) {
+            trackInfo->m_players.append(player->getGroup());               
+            trackAlreadyAdded = true;
+            break;
+        }
+    }
+    if (!trackAlreadyAdded) {
+        TrackInfo *newTrackInfo = new TrackInfo(pNewTrack);
+        newTrackInfo->m_players.append(player->getGroup());
+        m_trackList.append(newTrackInfo);                
+        connect(&m_trackList.last()->m_trackInfo,SIGNAL(readyToBeScrobbled(TrackPointer)),
+                &m_broadcaster,SLOT(slotReadyToBeScrobbled(TrackPointer)));
+    }
+    //A new track has been loaded so must unload old one.
+    resetTracks();
+}
+
+void ScrobblingManager::slotPlayerEmpty() {
+    QMutexLocker locker(&m_mutex);
+    resetTracks();
+}
+
+void ScrobblingManager::resetTracks() {
+    for (TrackToBeReset candidateTrack : m_tracksToBeReset) {
+        QLinkedList<TrackInfo*>::Iterator trackListIterator =
+            m_trackList.begin();
+        for (;trackListIterator != m_trackList.end(); ++trackListIterator) {
+            TrackInfo *trackInfo = *trackListIterator;
+            if (trackInfo->m_pTrack == candidateTrack.m_pTrack) {
+                if (!trackInfo->m_players.contains(candidateTrack.
+                    m_playerGroup)) {
+                    qDebug() << "Track doesn't contain player"
+                                "yet is requested for deletion.";
+                    break;
+                }
+                //Load error, stray from engine buffer.
+                BaseTrackPlayer *player = 
+                    m_pManager->getPlayer(candidateTrack.m_playerGroup);
+                if (player->getLoadedTrack() ==
+                    candidateTrack.m_pTrack) 
+                    break;                    
+                QLinkedList<QString>::iterator it = 
+                    trackInfo->m_players.begin();
+                while (it != trackInfo->m_players.end() &&
```

Comment:
3 nested loops with lots of conditional statements feels way too heavy for the task of synchronizing internal state. Please rethink this piece of code. It also contains the sensitive memory management as I mentioned out elsewhere.

Reply:
This is inefficient, yes, but the number of tracks is so low I haven't seen any performance hits. I didn't want to optimize unless I saw any so I think the code is fine this way, there will be at most 8 different tracks between both lists. Also I don' t understand what you mean with the sensitive memory management. I delete a pointer I created in the slotNewTrack. Since I know where the pointer is coming from I don't see any problem with it. I can wrap it in a unique_ptr though if you think it's better.

Reply:
I'm not complaining about efficiency, I'm talking about clean and readable code ;)

Reply:
If its not about efficiency, how would you go about iterating through the list of tracks to be reset, picking each track to be reset, looking for it on the list of current tracks and then deleting only the player that's been ejected from the list of players each tracked track has without 3 loops? I'm open to suggestions.